scalableminds/webknossos
# Custom segment colors with cuckoo hashing

### URL of deployed dev instance (used for testing):
- https://segmentcolorcuckoo.webknossos.xyz/

### Todo:
- [X] implement and adapt cuckoo table for storing colors
- [X] build UI to edit colors
- [X] use integer textures to store colors for 32 bit segment ids
- [X] refactor initial proof-of-concept
- [x] use custom colors for meshes, too
- [x] make compatible with multiple segmentation layers
- [x] make compatible with JSON mappings
  - [x] unify with custom colors in JSON mappings
- [X] fix conceptual bug in merger mode (random colors vs unique ids)
  - [x] diffing segment map should work against correct map
  - [X] hiding volume layer produces invalid shader
  - [x] fix that segment colors won't work when the tracing is loaded with no segmentation visible initially
- [x] doublecheck compatibility with texture count constraints
- [x] debounce update_segments actions or compact the update actions
- [x] test thoroughly

### Steps to test:
- open an annotation with a segmentation layer
  - click a segment
  - go to the segments tab and rightclick a segment to change its color
  - also load an ad-hoc mesh for that segment
  - change the color again --> the color in the viewport, in the segments tab and in the 3D viewport for the mesh should change
  - save & reload
  - the set color should still be there
- add a second volume layer
  - brush and change the segments color
  - switch between the volume layers and ensure that the colors are not mixed up somehow
- test the merger mode a bit
- open a dataset with a segmentation JSON mapping that has custom colors (e.g., ROI2017_wkw with `mini-mapping-with-colors.json`)
  - activate that mapping and see that the custom colors were used

### Issues:
- #6306 

------
(Please delete unneeded items, merge only when none are left open)
